pkg/alexa/gen/skill.go

```diff
@@ -6,415 +6,283 @@ import (
+	loc.Set(s.skillInstructions, []string{instructions})
+	return s
 }
 
-func (s *SkillBuilder) AddIntent(intent Intent) {
-	s.intents = append(s.intents, intent)
+func (s *SkillBuilder) WithPrivacyFlag(flag string, value bool) *SkillBuilder {
+	s.privacyFlags[flag] = value
+	return s
 }
 
-// AddIntentString creates an Intent from the string, adds it and returns a reference.
-func (s *SkillBuilder) AddIntentString(intent string) *Intent {
-	i := NewIntent(intent)
-	s.intents = append(s.intents, i)
-	return &i
-}
+func (s *SkillBuilder) AddLocale(locale string) *SkillLocaleBuilder {
+	s.registry.Register(l10n.NewLocale(locale))
 
-func (s *SkillBuilder) AddType(t Type) {
-	// TODO: validate rules!
-	s.types = append(s.types, t)
+	lb := NewSkillLocaleBuilder(locale).
+		WithLocaleRegistry(s.registry)
+	s.locales[locale] = lb
+	return lb
 }
 
-func (s *SkillBuilder) AddTypeString(t string) {
-	s.types = append(s.types, Type(t))
-}
-
-func (s *SkillBuilder) AddCountry(c alexa.Country) {
-	s.countries = append(s.countries, c)
+func (s *SkillBuilder) AddCountry(country string) *SkillBuilder {
+	s.countries = append(s.countries, country)
+	return s
 }
 
-func (s *SkillBuilder) AddCountries(cs []alexa.Country) {
+func (s *SkillBuilder) AddCountries(cs []string) *SkillBuilder {
 	for _, c := range cs {
 		s.countries = append(s.countries, c)
 	}
+	return s
 }
 
 // Build builds an alexa.Skill object.
-// TODO: return errors!
 func (s *SkillBuilder) Build() (*alexa.Skill, error) {
+	if s.registry == nil || len(s.registry.GetLocales()) == 0 {
+		return nil, fmt.Errorf("No locales registered to build")
+	}
+	//
+	if s.locales == nil || len(s.locales) == 0 {
+		for n, _ := range s.registry.GetLocales() {
+			s.locales[n] = NewSkillLocaleBuilder(n).
+				WithLocaleRegistry(s.registry)
+
+		}
+	}
+
+	dl := s.registry.GetDefault()
+
 	skill := &alexa.Skill{
 		Manifest: alexa.Manifest{
-			Version: "1.0",
+			Version:    "1.0",
+			Publishing: alexa.Publishing{},
 		},
 	}
-
-	// default is always set if at least one locale was defined.
-	dl, _ := s.locales[s.defaultLocale]
-
 	skill.Manifest.Publishing.Category = s.category
 	// TODO: ensure unique occurance
-	skill.Manifest.Publishing.Countries = s.countries
-
-	if s.testingInstructions != "" {
-		skill.Manifest.Publishing.TestingInstructions = s.testingInstructions
+	if len(s.countries) > 0 {
+		skill.Manifest.Publishing.Countries = s.countries
 	} else {
-		skill.Manifest.Publishing.TestingInstructions = dl.Translations.Get(l10n.KeySkillTestingInstructions)
+		skill.Manifest.Publishing.Worldwide = true
 	}
+	skill.Manifest.Publishing.TestingInstructions = dl.Get(s.skillInstructions)
 
-	// Permissions are required.
+	// TODO: Permissions are required.
 	skill.Manifest.Permissions = &[]alexa.Permission{}
 
 	// PrivacyAndCompliance is required.
 	skill.Manifest.Privacy = &alexa.Privacy{}
-	if s.privacy.flags[FlagIsExportCompliant] {
+	if s.privacyFlags[FlagIsExportCompliant] {
```

Comment:
This seems verbose. There must be a better way to do this. Perhaps let them add the `Privacy` struct to the builder.

Reply:
Also this move is influenced by "ease of use and readability" (autocompletion, ...). But additionally, if we would allow passing the `alexa.Privacy` it gets weird because we would build the `Locales` section into it when `Build()` is called, which would introduce unexpected "magic" IMHO ;)
Thinking further: when passing `alexa.Privacy` and the `With` function is used multiple times, do we merge the objects? Too much hassle IMHO compared to defining flag constants and making it overly verbose.
```
// current
SkillBuilder().WithPrivacyFlag(gen.FlagIsExportCompliant, true)
// pass alexa.Privacy object
SkillBuilder().WithPrivacy(alexa.Privacy{
  IsExportCompliant: true,
  // Locales: ... // what if you pass this along, would it be overwritten during Build()?
})
```

I'm open to any suggestion how to pass those flags in an simpler/better way.